### ISSUE TITLE
oscillations in the loss

### ISSUE THREAD

**Issue by nicolas-dufour** [mlfoundations/open_clip#822]
Hi!
I see in the loss for cc3m model that there are oscillations in the loss.
When using the same webdataset framework I also have oscillations as well. In my case it's more annoying because the oscillation amplitude is greater than the decrease of the loss per epoch.

Do you know what can be the reason of such behaviour?
From my experiments, it seems to be linked to the webdataset since a traditional dataloader don't suffer from such issues. In my case the period of the oscillation is of the number of steps per epoch (on cc12m).

Thanks for the help!

**Comment by rom1504** [comment]
Did you shuffle the dataset?

**Comment by nicolas-dufour** [comment]
Hey @rom1504 
Yes i'm shuffling shards and samples. I use the following settings that are the openclip defaults
```
        shard_shuffle_size=2000,
        shard_shuffle_initial=500,
        sample_shuffle_size=5000,
        sample_shuffle_initial=1000,
```
Thanks for the help!

**Comment by jbbqqf** [comment]
Hi! Triaging older issues — this one looks like expected behavior rather than a bug, based on what @rwightman said in a related thread.

Evidence:
- [@rwightman in #556 (2024-11-17)](https://github.com/mlfoundations/open_clip/issues/556#issuecomment-2480879699): *"AFAIK we've never seen loss spikes synched with epochs like that, oscillations ala #822 that seem to line up with epochs aren't uncommon ... but dont' appear to have a significant impact and are nowhere near as significant as this based all the runs various people associated with this project have done."* — that's a direct maintainer disposition on this specific issue, treating per-epoch oscillations as a known benign characteristic.
- In-thread, [@rom1504's only diagnostic question](https://github.com/mlfoundations/open_clip/issues/822) was whether shuffling was on; you confirmed shard_shuffle_size=2000 / sample_shuffle_size=5000 (open_clip defaults). After that the thread went quiet — no further actionable signal surfaced.
- Mechanistically this matches webdataset's shard-bucketed sampling: with finite shuffle buffers the same shards get revisited at epoch boundaries, so the per-step loss distribution differs slightly between intra-epoch batches and epoch-boundary batches. It's a property of the sampler, not the loss / training code.

If I'm wrong about your oscillation amplitude being in the same "no significant impact" regime that @rwightman is describing (e.g. you have repro evidence that it actively hurts downstream task performance), happy to dig further — otherwise would you mind closing this out?

---
*Disclosure: I drafted this comment with help from Claude Code while triaging stale issues; the cross-reference to #556 and the in-thread quotes were verified manually against the upstream comment IDs.*
